### DISCUSSION TITLE
How to use `#[auto_type]` with `as_select()`

### DISCUSSION THREAD

**Discussion by marcaddeo** [diesel-rs/diesel#4408] [Q&A] [answered]
I'm trying to use `#[auto_type]` with a function that returns a select statement, which uses `as_select()` but I can't seem to get it working. Am I doing something wrong here?

Example code:

```rust
#[derive(Debug, Default, Queryable, Identifiable, Selectable, Insertable, Associations)]
#[diesel(table_name = crate::schema::post)]
#[diesel(belongs_to(UserRecord, foreign_key = user_id))]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct PostRecord {
    pub id: i32,
    pub user_id: i32,
    pub content: String,
}

#[diesel::dsl::auto_type]
fn post_select() -> _ {
    post::table.select(PostRecord::as_select())
}
```

Which gives me an `ambiguous associated type` error that I can't seem to resolve.

On the other hand, this does work:

```rust
#[diesel::dsl::auto_type]
fn post_select() -> _ {
    post::table.select((post::id, post::user_id, post::content))
}
```

But obviously using `as_select()` is a bit more ergonomic.

This seems like it should work to me, but maybe I'm missing something.

**Comment by weiznich** [comment, marked as answer]
`Selectable::as_select` is one of the functions that is not supported by `#[auto_type]` so it is kind of expected to not work there. With the current limitations of what a proc-macro can do and with the current design of that function it's not possible to support it there as it has an unbounded generic type somehow needs to be specified. 


The [documentation already] (https://docs.diesel.rs/2.2.x/diesel/dsl/attr.auto_type.html#limitations ) states that `#[auto_type]` doesn't support all of diesels DSL, although this function is missing from the list in the documentation. We would welcome a PR to add it there. 

The workaround (provide an explicit type for this subexpression) listed in the documentation should also work here.

**Reply by marcaddeo** [reply]
Thanks! I was able to get it working as such:

```rust
#[diesel::dsl::auto_type]
fn post_select() -> _ {
    let as_select: AsSelect<PostRecord, Sqlite> = PostRecord::as_select();
    post::table.select((as_select,))
}
```

**Reply by weiznich** [reply]
Great to hear that you got it to work. I would advice to use `select(as_select)` instead of the nested tuple as the later can have interesting effects on deserialisation.
